### ISSUE TITLE
Yabai behaving strangely with Arc Browser

### ISSUE THREAD

**Issue by GonzaloRizzo** [koekeishiya/yabai#1622]
I'm giving [Arc Browser](https://arc.net/) a try but it's PIP feature doesn't work properly when Yabai is enabled.
If I try clicking the buttons my cursor get's moved somewhere else and the same happens if I try to drag the window.

I tried using yabai rules so that window is not managed but that doesn't seem to work either for some reason.

This is what I get when I query the windows.
```json
[
  {
    "id": 10021,
    "pid": 75372,
    "app": "Arc",
    "title": "",
    "frame": {
      "x": 1190,
      "y": 55,
      "w": 600,
      "h": 337
    },
    "role": "AXWindow",
    "subrole": "AXSystemDialog",
    "display": 1,
    "space": 1,
    "level": 3,
    "opacity": 1,
    "split-type": "none",
    "split-child": "none",
    "stack-index": 0,
    "can-move": true,
    "can-resize": true,
    "has-focus": false,
    "has-shadow": true,
    "has-border": false,
    "has-parent-zoom": false,
    "has-fullscreen-zoom": false,
    "is-native-fullscreen": false,
    "is-visible": true,
    "is-minimized": false,
    "is-hidden": false,
    "is-floating": true,
    "is-sticky": false,
    "is-topmost": true,
    "is-grabbed": false
  },
  {
    "id": 8258,
    "pid": 75372,
    "app": "Arc",
    "title": "New Issue · koekeishiya/yabai",
    "frame": {
      "x": 0,
      "y": 44,
      "w": 1800,
      "h": 1070
    },
    "role": "AXWindow",
    "subrole": "AXStandardWindow",
    "display": 1,
    "space": 1,
    "level": 0,
    "opacity": 1,
    "split-type": "none",
    "split-child": "none",
    "stack-index": 0,
    "can-move": true,
    "can-resize": true,
    "has-focus": true,
    "has-shadow": true,
    "has-border": true,
    "has-parent-zoom": false,
    "has-fullscreen-zoom": false,
    "is-native-fullscreen": false,
    "is-visible": true,
    "is-minimized": false,
    "is-hidden": false,
    "is-floating": true,
    "is-sticky": false,
    "is-topmost": false,
    "is-grabbed": false
  }
]
```

Based on this I made this rule `yabai -m rule --add app="^Arc$" subrole="^AXSystemDialog$" manage=off` but it doesn't seem to make an effect.
I also tried to disable Arc altogether by doing `yabai -m rule --add app="^Arc$" manage=off` and the main Arc windows gets unmanaged but the PIP one somehow still is.

Any clues?

**Comment by ChronoStriker1** [comment]
i have this in mine for arc and it seems to work correctly.  I put the title in as "" since that window doesnt seem to have one, so it only seems to affect it.

yabai -m rule --add app="^Arc$" title="" manage=off

**Comment by GonzaloRizzo** [comment]
That's weird, I tried that but it says `invalid key-value pair 'title='`, tried using title="^$"` instead of `title=""` but I still can't click the buttons on the PIP window and dragging still doesn't work well

**Comment by BeeMargarida** [comment]
I have similar problems, even more when trying to adjust the pop-up player when I switch tabs, the focus is lost and my mouse is moved to an edge of my screen. Any idea how to possibly solve this? Thank you!

**Comment by gattlegun** [comment]
Using the following rule fixed this issue for me:
```bash
yabai -m rule --add app="^Arc$" title="^$" mouse_follows_focus=off
```

**Comment by BeeMargarida** [comment]
> Using `mouse_follows_focus` instead of `manage` fixes this issue for me:
> 
> ```shell
> yabai -m rule --add app="^Arc$" title="^$" mouse_follows_focus=off
> ```

Seems to be working, thank you 🙌

**Comment by aaronrolls** [comment]
This is what works for me:

`yabai -m rule --add app="Arc.LittleBrowserWindow" manage=off`

That should make the little browser window unmanaged while the main app remains managed.

**Comment by josekoalas** [comment]
> Using `mouse_follows_focus` instead of `manage` fixes this issue for me:
> 
> ```shell
> yabai -m rule --add app="^Arc$" title="^$" mouse_follows_focus=off
> ```

This works too with Orion browser's popups changing Arc to Orion if anyone is wondering. Thank you so much for the solution!

**Comment by gilitos92** [comment]
> Using `mouse_follows_focus` instead of `manage` fixes this issue for me:
> 
> ```shell
> yabai -m rule --add app="^Arc$" title="^$" mouse_follows_focus=off
> ```

This fixes the issue with the mini player and PIP.

To fix the issue of Yabai trying to manage the Little Arc window this works:
```
yabai -m rule --add app="^Arc$" title="^Space 1$" manage=off
```

**Comment by teelahti** [comment]
Tried all the above solutions and none are really working as intended: Little Arc floats on top of the other windows, but Yabai makes space for it (which stays empty and desktop is shown on that place).

**Comment by popbones** [comment]
For what it's worth I think Arc has changed the window properties for Little Arc. So you may want to run a window query first to see what properties are actually on the Little Arc pop-up first by doing:

```
yabai -m query --windows
```

I'm using Version 0.93.0 (37600) today. Arc set the title of the Little Arc pop-up to "Space 1" for me. So @gilitos92 's answer worked for me.

> > Using `mouse_follows_focus` instead of `manage` fixes this issue for me:
> > ```shell
> > yabai -m rule --add app="^Arc$" title="^$" mouse_follows_focus=off
> > ```
> 
> This fixes the issue with the mini player and PIP.
> 
> To fix the issue of Yabai trying to manage the Little Arc window this works:
> 
> ```
> yabai -m rule --add app="^Arc$" title="^Space 1$" manage=off
> ```

Here is the role I'm using for now:

```
yabai -m rule --add app="^Arc$" title="^Space [0-9]+" manage=off
```

**Comment by gilitos92** [comment]
> For what it's worth I think Arc has changed the window properties for Little Arc. So you may want to run a window query first to see what properties are actually on the Little Arc pop-up first by doing:
> 
> ```
> yabai -m query --windows
> ```
> 
> I'm using Version 0.93.0 (37600) today. Arc set the title of the Little Arc pop-up to "Space 1" for me. So @gilitos92 's answer worked for me.
> 
> > > Using `mouse_follows_focus` instead of `manage` fixes this issue for me:
> > > ```shell
> > > yabai -m rule --add app="^Arc$" title="^$" mouse_follows_focus=off
> > > ```
> > 
> > 
> > This fixes the issue with the mini player and PIP.
> > To fix the issue of Yabai trying to manage the Little Arc window this works:
> > ```
> > yabai -m rule --add app="^Arc$" title="^Space 1$" manage=off
> > ```
> 
> Here is the role I'm using for now:
> 
> ```
> yabai -m rule --add app="^Arc$" title="^Space [0-9]+" manage=off
> ```

Nice! Didn't though of covering all "Space X" numbers. Is there any reason why you are using it?

Btw, another way of obtaining window properties is using the "Accessibility Inspector" App that comes with Xcode.

**Comment by popbones** [comment]
> > For what it's worth I think Arc has changed the window properties for Little Arc. So you may want to run a window query first to see what properties are actually on the Little Arc pop-up first by doing:
> > ```
> > yabai -m query --windows
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm using Version 0.93.0 (37600) today. Arc set the title of the Little Arc pop-up to "Space 1" for me. So @gilitos92 's answer worked for me.
> > > > Using `mouse_follows_focus` instead of `manage` fixes this issue for me:
> > > > ```shell
> > > > yabai -m rule --add app="^Arc$" title="^$" mouse_follows_focus=off
> > > > ```
> > > 
> > > 
> > > This fixes the issue with the mini player and PIP.
> > > To fix the issue of Yabai trying to manage the Little Arc window this works:
> > > ```
> > > yabai -m rule --add app="^Arc$" title="^Space 1$" manage=off
> > > ```
> > 
> > 
> > Here is the role I'm using for now:
> > ```
> > yabai -m rule --add app="^Arc$" title="^Space [0-9]+" manage=off
> > ```
> 
> Nice! Didn't though of covering all "Space X" numbers. Is there any reason why you are using it?
> 
> Btw, another way of obtaining window properties is using the "Accessibility Inspector" App that comes with Xcode.

Thanks for the advice on using Accessibility Inspector. I used a wider pattern just because I do actually have multiple Spaces configured in my Arc browser. However, I have no clue whether or not Arc would actually use titles other than "Space 1", but it doesn't hurt to anticipate if there's 1, there might be 2.

**Comment by gilitos92** [comment]
Hi! Just wanted to update everyone that after updating to Arc version 0.96.0 (38004), the properties of both Little Arc and PiP windows have apparently changed to "AXSystemDialog". As a result, all previous rules like "title='Space 1'" are now useless. However, this is good news because now a single rule can fix both problems at once:

```
yabai -m rule --add app="^Arc$" subrole='AXSystemDialog' manage=off mouse_follows_focus=off
```

**Comment by gilitos92** [comment]
Damn, you are right, I think my brain slipped a little there...

**Comment by wojciech-kulik** [comment]
This one works for me with Litte Arc:
```
yabai -m rule --add app='^Arc$' title='^Space.*$' manage=off
```

**Comment by joshstobbs** [comment]
None of these seem to be working for me - does anyone have ideas on if something has changed?

**Comment by sebb3** [comment]
Finally found a hacky workaround for this...

The only mention of "Little Arc" I found was in the window menu, there it shows up as "Little Arc - {title}".

You can use it like this:
Add a signal for the window_created event with app="Arc". Then you can utilize an apple script that checks if the title of the window is part of any menu item under Windows that starts with Little Arc. If there is a match you can float the window. (Or go the opposite route: Always float Arc windows, and tile them if it does NOT match "Little Arc")

**Comment by ognistik** [comment]
For dealing with Arc and Little Arc I have a solution that includes using [OpenIn](https://loshadki.app/openin4/) and [a Shortcut](https://www.icloud.com/shortcuts/3beff1ba091746a58bf15f7575792847). There’s four scenarios that will open Little Arc window, let me give you some screen recordings for you to check Yabai's behavior on my end:  
1. When Little Arc is triggered within Arc clicking a link with CMD + OPT.

https://github.com/koekeishiya/yabai/assets/124819548/f096d6df-4e9f-47a1-af9f-ac3bfbea9ea8

2. When a new Little Arc is created with CMD + OPT + CTRL + N (with Arc in the background).

https://github.com/koekeishiya/yabai/assets/124819548/73b44a12-e048-49b2-abf0-0186763ed773

3. When Arc is already running  (in the background or not) and a link is clicked from another app.

https://github.com/koekeishiya/yabai/assets/124819548/1d182e89-425f-4211-9445-6967a33c0ab9

4. When Arc is closed and a link clicked from another app.

https://github.com/koekeishiya/yabai/assets/124819548/0c82a2b3-13ae-4404-8d2e-be0248e1556e

**For scenarios 1 and 2, the following rule seems to still be effective:**

```
yabai -m rule --add label="Little Arc" app="^Arc$" title="^$" manage=off layer=below
``` 

Notice I am using `layer=below`  to deal with the weird auto behavior of layers mentioned [HERE](https://github.com/koekeishiya/yabai/issues/1929). You may remove it if you don’t need it. 

**For scenarios 3 and 4 I found the best way to deal with Little Arc is by intercepting links BEFORE they are sent to Arc** and add a signal that tells Yabai to open the next window as float. This is the only workaround that’s worked for me, at least until Arc decides to add something to the Little Arc window title that differentiates it. I have set this up in [OpenIn](https://loshadki.app/openin4/) (which I was using anyways to set rules for links to open in different browsers). 

So, you set OpenIn as default browser. Then add the Shortcuts app as if it was a browser. You add a URL replacement for Regex to replace` https://(.*)`  with `shortcuts://run-shortcut?name=LittleArcYabai&input=https%3A%2F%2F$1`  which opens my [LittleArcYabai Shortcut](https://www.icloud.com/shortcuts/3beff1ba091746a58bf15f7575792847) with whatever URL as input. I set this as default for any link (or you can choose rules for it). Here’s a visual guide:

1. [OpenIn Interface and Overview](https://github.com/koekeishiya/yabai/assets/124819548/db93e50d-fe38-43f6-8c3e-3a40939df027)
2. [The Regex Replacement](https://github.com/koekeishiya/yabai/assets/124819548/4673ccae-6dfc-4d02-a681-5d9994e5742e)
3. [The Rule](https://github.com/koekeishiya/yabai/assets/124819548/b5967bac-3a54-46c5-8c84-3463c4a93578)

The Shortcut first runs a Shell Script which checks if Arc is already running or not. Then:
1. IF Arc is already running, the script adds a signal for Yabai to open the following window as Float. Then it opens the passed URL with Applescript in Little Arc.
2. IF Arc is NOT running, then Arc main window will open (tiled in Yabai) with the link in a normal window. I set it up this way because I’ve always been bothered by the empty main window opening and then another Little Arc aside from it. The double window behavior messes up with the script that adds a signal to open the next window as float, so better avoid Little Arc altogether for this first opening of the app.

You can download the Shortcut above, or here’s the code I use:

```
if pgrep -x "Arc" > /dev/null
then
/opt/homebrew/bin/yabai -m signal --add label=float_next_window_created event=window_created action='/opt/homebrew/bin/yabai -m signal --remove float_next_window_created; /opt/homebrew/bin/yabai -m signal --remove float_next_application_launched; /opt/homebrew/bin/yabai -m window $YABAI_WINDOW_ID --toggle float --layer below'

    echo "OpenAsLittleArc"

else
    echo "OpenAsNormal"
fi
``` 

The AppleScript to open LittleArc
```
on run {input, parameters}
	tell application "Arc" to open location (item 1 of input)
end run
``` 

The AppleScript to open in main window if Arc is closed
```
on run {input, parameters}
	tell application "System Events"
		if not (exists process "Arc") then
			-- Arc is not running, so open the application
			tell application "Arc" to activate
			delay 0.5 -- Wait for Arc to launch
		end if
	end tell
	tell application "Arc"
		set windowCount to count of windows
		if windowCount is 0 then
			-- No open windows, create a new window
			activate
			tell application "System Events"
				keystroke "n" using command down -- Press CMD + N
				delay 1
			end tell
		end if
		tell front window
			tell space 1
				make new tab with properties {URL:input}
			end tell
		end tell
	end tell
end run
``` 

[HERE's a visual of the Shortcut.](https://github.com/koekeishiya/yabai/assets/124819548/727af2da-e476-4496-8fa9-ed805eb5c1df)

This is, of course, a hacky workaround and far from ideal, but it makes Arc and Yabai usable together for me. The downside is that you have to keep the Shortcuts app running in the background (even without any windows) because that’s the only way that it can trigger scripts (or so it seems). I’ve tried to do this with Keyboard Maestro instead of Shortcuts but OpenIn doesn’t see to work well with it.

**Comment by koekeishiya** [comment]
Think this is fixed in v6.0.3.
